[PM-1796] The autofill keyboard shortcut does not prompt a user to unlock a locked extension within an incognito browsing session

cagonzalezcs opened the pull request:
## Type of change

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Users who attempt to enact an autofill with a locked extension within an incognito browsing session are not being prompted to unlock the extension using their master password. The reason why this is happening is due to how we're attempting to open a new tab within the incognito session. Chrome blocks the ability to open a tab within a inactive window, and due to how Chrome treats the execution of extension logic, an incognito window will always be considered "inactive".

To mitigate this issue, we are removing the passed sender tab that is provided when attempting to call the `BrowserApis.openBitwardenEtensionTab()`. We then focus the window where the tab was recently opened to direct user attention to that window (goes away from the incognito window). Once the user password is inserted, we focus the incognito window and re-attempt to autofill the password.

Unfortunately, this redirection of the user focus is required in this instance, something which would be better to avoid. However, it hopefully will ensure that users are not attempting to guess where the should be attempting to login. It'd be ideal to keep the login tab in the same window as the incognito window, but this is a decent alternative. 

Note: It is likely that this issue is related to similar behavior experienced within https://github.com/bitwarden/clients/issues/2243 and the changes in this PR should resolve https://github.com/bitwarden/clients/issues/2243. Also, after some deliberation a fix has been provided that should resolve https://github.com/bitwarden/clients/issues/5115 and closes https://github.com/bitwarden/clients/pull/5116 as it will no longer be required.

## Code changes

- **apps/browser/src/browser/browserApi.ts:** These changes modify a couple of the methods within the `BrowserApi` class, and introduce a new method to the class to allow for focusing a window based on tab id. First we are modifying `BrowserApi.createNewTab` to contain a returned promise which will resolve to the tab that was recently created by the `chrome.tabs.create` api call. This modification was made with an effort to not modify how the method is called directly, which would necessitate a larger refactor. However, it does allow us to pull the tab information if necessary using an await promise. We've also modified the method to remove the `openerTab` option as it is no longer being used in this context.  The second change made is to the `BrowserApi.openBitwardenExtensionTab` method to anticipate an attempt to open the tab during an incognito session. If that does occur, we open the tab without passing the original tab sender info and then focus the window where the newly created tab exists. Otherwise, we create the new tab as previously expected while passing the sender info. We've also created a method has been added named `BrowserApi.focusWindow` which simply accepts a window id, and calls `chrome.windows.update` to focus the window with that id value. Finally, we've removed the attempt to focus the tab when closing the Bitwarden Login form, as this will be handled when the `unlocked` event is called.

- **apps/browser/src/background/runtime.background.ts:** A small modification was made to how we process the `unlocked` and `loggedIn` extension messages. We first focus the window id before attempting to retry a failed command. After addressing some fixes for Firefox, we also now set the tab where the failed command is being tried as an active tab shortly after focusing the window.

## Screenshots

https://user-images.githubusercontent.com/16629865/235764985-d0132a04-3bbf-436e-9217-309638a85cb9.mov

https://user-images.githubusercontent.com/16629865/236022560-864a8ebf-66a5-4771-ac8f-3909d372c427.mov

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

djsmith85 commented:
Requesting review from @eliykat because of his previous work on https://github.com/bitwarden/clients/pull/4808

eliykat commented:
Thanks for the mention. @cagonzalezcs, this overlaps with #5116 which actually removes the use of the `openerTab` parameter, because it appears to be incompatible with Firefox containers. I've eyeballed that code but haven't tested it yet as I've been busy with Admin Console priorities. Given that both PRs are related and fall within autofill's code ownership, do you mind taking over that PR review so you can ensure they're consistent? I'm happy to review or assist in any way, or let me know if you don't have capacity for that and we can figure something out.

cagonzalezcs commented:
@eliykat @djsmith85 

I've found a path forward on more conclusively addressing the issues identified within https://github.com/bitwarden/clients/issues/5115 

Essentially, we can indeed remove the reference to the `openerTabId` when calling `chrome.tabs.create`, which is what was causing the issues in both Chrome incognito mode as well as when using Firefox with the Multi-Account Containers addon. 

With this removal, we still need to focus the expected tab when attempting to re-try autofill. To do this, once we have triggered a `loggedIn` or `unlocked` message, if we are retrying a command, we first focus the window and set active the tab where that action is being retried. The behavior then appears as shown in the following video:

https://user-images.githubusercontent.com/16629865/236022560-864a8ebf-66a5-4771-ac8f-3909d372c427.mov

The changes introduced with this comment maintain the behavior addressed for incognito windows.

djsmith85 commented on apps/browser/src/browser/browserApi.ts:
```diff
@@ -127,24 +127,31 @@ export class BrowserApi {
+  static async openBitwardenExtensionTab(
     relativeUrl: string,
     active = true,
     openerTab?: chrome.tabs.Tab
```
If we don't require this anymore, we should remove it.

cagonzalezcs replied:
@djsmith85 

We are using this on line 153 of the file to await the creation of the tab before focusing the window. 

```
    const createdTab = await this.createNewTab(url, active);
    this.focusWindow(createdTab.windowId);
```

I can set this up as a `.then` chain if it's preferable, but my personal preference is to leverage async/await for most promise handling.

Jet let me know if you'd like to see the change.

cagonzalezcs replied:
@djsmith85

We are using this on line 153 of the file to await the creation of the tab before focusing the window.

```
    const createdTab = await this.createNewTab(url, active);
    this.focusWindow(createdTab.windowId);
```

I can set this up as a .then chain if it's preferable, but my personal preference is to leverage async/await for most promise handling.

Jet let me know if you'd like to see the change.

cagonzalezcs replied:
I did find, however, that we could remove the `async` declaration I added to `createNewTab` since we weren't really awaiting anything within the function call :)

djsmith85 replied:
I was referring to the `openerTab`-param

cagonzalezcs replied:
@djsmith85 

Ah, I misunderstood. Good call, I've gone ahead and removed that parameter and have updated any calls to the method that passed that value to the function call.

djsmith85 reviewed (changes requested):
Changes are looking good and my limited testing (unlock on autofill, Save/change password prompt including new edit functionality) also checks out.

Nice work, just a minor nit on the `openBitwardenExtensionTab`-method

eliykat commented on apps/browser/src/browser/browserApi.ts:
```diff
@@ -127,24 +127,27 @@ export class BrowserApi {
-  static createNewTab(url: string, active = true, openerTab?: chrome.tabs.Tab) {
-    chrome.tabs.create({ url: url, active: active, openerTabId: openerTab?.id });
+  static createNewTab(url: string, active = true): Promise<chrome.tabs.Tab> {
+    return new Promise((resolve) =>
+      chrome.tabs.create({ url: url, active: active }, (tab) => resolve(tab))
+    );
```
Nice solution for getting the tab reference out of the callback.

eliykat reviewed (approved):
Nice one, thanks for handling the Firefox issue at the same time.

djsmith85 reviewed (approved):
LGTM
